### DISCUSSION TITLE
How to analyze dmrg wfn?

### DISCUSSION THREAD

**Discussion by jamcas** [pyscf/pyscf#2643] [Q&A]
1.How to get CI components in dmrgscf , that's very important in complex system with large active space to distinguish which state i got.
2.Can i get spin-related properties such as local spin and spin correlation **(**_Nature Chemistry volume 6, pages 927–933 (2014)_ spin correlation function is (<S1*S2>)**)**  in dmrg?I know local spin can be implemented in CASSCF, but there are some problems in dmrg, which may be caused by mc.ci, which is ndarray in CASSCF and 0 in DMRGSCF.

**Comment by gkc1000** [comment]
1. Hi, getting CI components is not currently supported out of the box. There is unsupported code to sample CI coefficients from the DMRG wavefunction computed by Block. We can make this available if necessary.

2. Spin related properties can be obtained by processing the 2-particle DM computed in DMRG. You  need to write a small script to do so. For example, Si.Sj = Sz(i)Sz(j) + 1/2(S+(i)S-(j)) + S-(i)+S+(j))
and Sz = 1/2(a+i_up a_i_up - a+i_down a_i_down) etc. so you see that Si.Sj can be read out by a simple processing of the 2PDM dumped out by the DMRG program.

**Comment by jamcas** [comment]
HI, @gkc1000 , thanks very much for your instruction and consideration. In my opinion, getting CI components is very necessary, becasue it can help us find the right electronic structure especially some large complex systems, such as iron-sulfur clusters and nitrogenases, to help understand their reaction mechanism. It would be great if pyscf can make this avaliable.

As we know,  we can get 1-particle DM(dm1a+dm1b) , 2-particle DM(dm2) by make_rdm12 and dm1a,dm1b by make_rdm1s,If I understand correctly, the formula is：
```
<Si.Sj> = 0.25(dm1ai - dm1bi)(dm1aj - dm1bj) + 0.5*(S+(i)S-(j)+S-(i)S+(j))
```
But I don't know how to get S+(i)S-(j)  and S+(j)S-(i)  from dm2([aaaa]+[aabb]+[bbaa]+[bbbb]). Would  it be possible to give me some tips?

**Comment by jamcas** [comment]
I found https://github.com/pyscf/pyscf/blob/master/pyscf/fci/spin_op.py has the formula to compute S+S-and S-S+, but this fromula requires dm2ab which cannot be obtained directly in dmrg. In dmrg, we can only get dm2(dm2aa+dm2ab+dm2ba+dm2bb)

**Comment by zhendongli2008** [comment]
Hi there,

The function szHS(onepdm,twopdm,n,s) in the following script can compute
the spin density matrix
from spin-free 2RDM output from block,
https://github.com/zhendongli2008/autoGenInput/blob/master/pclusterN/spinAnalysis/spinTools.py
Here, n is the no. of electrons and s is the total spin quantum number. The
arrays
onepdm and twopdm can be computed as follows:

twopdm = 2.0*rdm2

onepdm = numpy.einsum('ijjl->il',rdm2)/(n-1)*2.0

where rdm2 is the spin-free 2RDM from BLOCK. A full example can be found
from here:
https://github.com/zhendongli2008/autoGenInput/blob/master/pclusterN/spinAnalysis/spinSpinCorr.py

Best,
Zhendong



jamcas <notifications@github.com> 于2019年7月2日周二 下午3:05写道：

**Comment by jamcas** [comment]
Thank you very much for providing such a powerful spin analysis tool！
I use spinSpinCorr.genCorr function to analyze a [4Fe-4S] cluster but it produced a strange result:
```
. . .
. . .
sisj:
[[ 1.26958891 -0.55095219 -0.16709273 -0.44084202]
 [-0.55095219  1.79702936 -0.41468235 -0.41316507]
 [-0.16709273 -0.41468235  0.99504119 -0.30341897]
 [-0.44084202 -0.41316507 -0.30341897  1.26864722]]
. . .
. . .
```
Referring to our dft result, in our [4Fe-4S] cluster, there should be two faces, and the spin correlation function of the two irons on each face should be positive. However,in the above result, none of sisj is positive. My script is as follows：
```
Fe = [None]*4
# These orbitals are selected after the natural orbit of boys localization
Fe[0] = [1, 4, 7, 15, 16]  
Fe[1] = [3, 9, 11, 10, 19]
Fe[2] = [2, 6, 13, 14, 18]
Fe[3] = [0, 5, 8, 12, 17]
groups = [Fe[0], Fe[1], Fe[2], Fe[3]]
ng = len(groups)
n = 23
s = 0.5

spinSpinCorr.loadRDMs('xxx/node0/spatial_twopdm.0.0.txt', 'a')
spinSpinCorr.genCorr('a.npy', 'ss.h5')
```
May be my usuage is wrong.Would it be possible to give me some tips about this.? If you have any questions or concerns, don’t hesitate to let me know.
  
Thanks you, and I looking forward to hearing from you.
                                                                                                                                                     
Best wishes 
jamcas

**Comment by zhendongli2008** [comment]
Hi Jamcas,

I think your usage of the script seems to be correct. However, the
inconsistency of your results
with DFT may be caused by several reasons. It looks like the local states
of irons are quite different
according to the diagonal parts of sisj. I'm not sure about the details of
your calculations, e.g.,
what type of your calculation (SCF? CASCI? CASSCF?) was and whether it was
performed correctly.

Best,
Zhendong

jamcas <notifications@github.com> 于2019年7月4日周四 下午4:33写道：

**Comment by jamcas** [comment]
This my script:
```
import os
from pyscf import gto, scf, mcscf, dmrgscf, lib, mrpt, qmmm
import sys
from pyscf.tools import molden
from pyscf import dft
import numpy

dmrgscf.settings.MPIPREFIX = 'mpirun -np 10'

os.system('date')

mol = gto.Mole()
mol.verbose = 4
mol.atom = open('baba-rc.xyz').read()
mol.basis = 'Def2-SVP'
mol.spin = 1
mol.charge = -2
mol.max_memory = 60000
mol.build()

# dft
mf = dft.UKS(mol)
mf.max_cycle = 500
mf.xc = 'b3lyp'
mf.level_shift = 0.5
mf.__dict__.update(lib.chkfile.load('baba-rc-scf.chk', 'scf'))

# dmrgscf
mc = dmrgscf.DMRGSCF(mf, 20, 23)
mc.max_cycle_macro = 1000
mc.conv_tol = 1e-4
mc.conv_tol_grad = 0.1
mc.fcisolver.memory = 60
mc.fcisolver.spin = 1
# add point charges
'''coords part'''
with open('baba-rc-coords') as f:
        coords = f.read()
coords = coords.split()
coords = numpy.array(coords)
coords = coords.astype('float64')
coords = coords.reshape(31377,3)

'''charges part'''
with open('baba-rc-charges') as f:
        charges = f.read()
charges = charges.split()
charges = numpy.array(charges)
charges = charges.astype('float64')
mc = qmmm.mm_charge(mc, coords, charges)
# define cas space\
mo = lib.chkfile.load('baba-rc-omega-point_charges-20o23e-guess-sort_mo-1.chk', 'mcscf/mo_coeff')
mc.run(mo)

mo_coeff_nature, ci_nature, occ_nature = mc.cas_natorb(sort=True)
mcscf.chkfile.dump_mcscf(mc, mo_coeff=mo_coeff_nature, mo_occ=occ_nature)
molden.from_mo(mol, 'baba-rc-omega-point_charges-20o23e-guess-sort_mo-1-check.molden', mo_coeff=mo_coeff_nature, ene=None, occ=occ_nature)

mc.analyze()

os.system('date')
```
The active space consists of 3d orbitals of four irons.

best wishes
jamcas

**Comment by zhendongli2008** [comment]
It is highly possible that after CASSCF your active orbitals are not longer
the target ones you desired.
You need to check the converged MOs. If they are not localized 3d, then it
is possible to have the strange results you got.

Best,
Zhendong


jamcas <notifications@github.com> 于2019年7月5日周五 下午2:56写道：

**Comment by jamcas** [comment]
Thank you very much for your tips.You said it is very correct.
I checked the MOs, they are indeed delocalized 3d orbitals and  I tried to convert the 3d natural orbitals in the active space into a localized orbit and then do dmrgscf but still got the same result, I think my usage is wrong:

```
# dmrgscf
. . . . . .
mo = lib.chkfile.load('baba-rc-omega-point_charges-20o23e-guess-sort_mo-1.chk', 'mcscf/mo_coeff')
mc.run(mo)
pz_idx = numpy.array([217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236])-1
loc_orb = lo.Boys(mol, mo[:,pz_idx]).kernel()
mo[:,pz_idx] = loc_orb
mc.run(mo)
. . . . . .
```
I noticed that you localized the orbits of the active space in your latest published article(_J. Chem. Phys. 150, 024302 (2019_).  I would be grateful if you could give me some instructions.

best wishes
jamcas

**Comment by zhendongli2008** [comment]
Sorry, I think I cannot help too much, since I don't know the details about
the composition of the active space.
My experience is that usually if you include Fe3d and S3p (bonding)
correctly, which is the minimal chemically
meaningful active space, CASSCF will not alter your results too much. Even
there can be some mixings between
these orbitals after calculations, you could relocalize them and transform
the RDMs to this localized basis
for analysis purpose.

Best,
Zhendong

>

**Comment by jamcas** [comment]
Thank you. I am sorry to have bothered you.You have helped me a lot, 
As you said, i should relocalize them and transform the RDMs to this localized basis. I think the RDMs is still delocalized, even if i use localized orbitals as active orbitals. So the key point is to  transform
the RDMs to this localized basis for the further analysis purpose. However, i don't know how to achieve this transformation. Can pyscf do this? Or I need write a script.
Thanks you, and I looking forward to hearing from you.
best wishes
jamcas

**Comment by sunqm** [comment]
<S1 S2> would need the 2-particle spin density matrices from the underlying dmrg calculations. I opened an issue in Block code https://github.com/sanshar/Block/issues/53. When the feature is ready in Block code, it should be pretty straightforward to enable it in pyscf

**Comment by davidsousarj** [comment]
> 1. Hi, getting CI components is not currently supported out of the box. There is unsupported code to sample CI coefficients from the DMRG wavefunction computed by Block. We can make this available if necessary.

HI, @gkc1000 . Is it possible to make this available for me? I will need it for my calculations.

**Comment by gkc1000** [comment]
This is easy to do with block2. Please post to the block2 github issues, or
read the block2 documentation.

**Comment by davidsousarj** [comment]
> This is easy to do with block2. Please post to the block2 github issues, or read the block2 documentation.

@gkc1000 i am using Block 1.5 which is, as far as I know, the latest version. Is block2 another package?

**Comment by fishjojo** [comment]
@davidsousarj https://github.com/block-hczhai/block2-preview

**Comment by verena-neufeld** [comment]
@jamcas , @davidsousarj , thanks for your questions. Has this been resolved?

**Comment by davidsousarj** [comment]
Hi @zhendongli2008 . Could you send me any references about the equations you used to calculate spin properties from the spin-free DMs in `spinTools.py`?

**Comment by bogdanoff** [comment]
I'm not sure it is exactly the same, but we have (re)derived relations for spin-dependent properties from spin-free RDMs in [10.1021/acs.jctc.1c00589](https://pubs.acs.org/doi/10.1021/acs.jctc.1c00589). There is also `mcscf.addons.make_spin_casdm1`  function following eq. 3 of [10.1002/qua.22320](https://onlinelibrary.wiley.com/doi/10.1002/qua.22320) to compute spin-resolved 1rdm from spin-free 2rdm.

**Comment by davidsousarj** [comment]
@bogdanoff thank you so much! The appendices in both papers look very useful to me.
